brainiak/reprsimil/brsa.py

```diff
@@ -867,38 +1456,264 @@ def _fit_RSA_UV(self, X, Y, X0,
+        # Prior parmeters for X0: T_X is transitioning matrix, Var_X
+        # is the marginal variance of the first time point. Var_dX is the
+        # variance of the updating noise.
+        sigma2_e = sigma_e ** 2
+        # variance of voxel-specific updating noise component
+        scan_onsets = np.setdiff1d(scan_onsets, n_T).astype(int)
+        n_scan = scan_onsets.size
+        total_log_p = 0
+        for scan, onset in enumerate(scan_onsets):
+            # Forward step
+            if scan == n_scan - 1:
+                offset = n_T
+            else:
+                offset = scan_onsets[scan + 1]
+            _, _, _, log_p_data, _, _, _, _, _ = \
+                self._forward_step(
+                    Y[onset:offset, :], T_X, Var_X, Var_dX, rho_e, sigma2_e,
+                    beta0)
+            total_log_p += log_p_data
+        return total_log_p
+
+    def _est_AR1(self, x, same_para=False):
+        """ Estimate the AR(1) parameters of input x.
+            Each column of x is assumed as independent from other columns,
+            and each column is treated as an AR(1) process.
+            If same_para is set as True, then all columns of x
+            are concatenated and a single set of AR(1) parameters
+            is estimated. Strictly speaking the breaking point
+            between each concatenated column should be considered.
+            But for long time series, this is ignored.
+        """
+        if same_para:
+            n_c = x.shape[1]
+            x = np.reshape(x, x.size, order='F')
+            rho, sigma2 = alg.AR_est_YW(x, 1)
+            # We concatenate all the design matrix to estimate common AR(1)
+            # parameters. This creates some bias because the end of one column
+            # and the beginning of the next column of the design matrix are
+            # treated as consecutive samples.
+            rho = np.ones(n_c) * rho
+            sigma2 = np.ones(n_c) * sigma2
         else:
-            return est_cov_AR1_UV, estU_chlsk_l_AR1_UV, est_SNR_AR1_UV, \
-                est_beta_AR1_UV, est_beta0_AR1_UV, \
-                est_sigma_AR1_UV, est_rho1_AR1_UV
+            rho = np.zeros(np.shape(x)[1])
+            sigma2 = np.zeros(np.shape(x)[1])
+            for c in np.arange(np.shape(x)[1]):
+                rho[c], sigma2[c] = alg.AR_est_YW(x[:, c], 1)
+        return rho, sigma2
+
+    def _forward_step(self, Y, T_X, Var_X, Var_dX, rho_e, sigma2_e, weight):
+        """ forward step for HMM """
+        # cov_form='diag'  We currently only implement diagonal form
+        # of covariance matrix for Var_X, Var_dX and T_X, which means
+        # each dimension of X is independent and their refreshing noise
+        # are also independent. Note that log_p_data takes this assumption.
+        if Var_X.ndim == 1:
+            inv_Var_X = np.diag(1 / Var_X)
+            half_log_det_Var_X = np.sum(np.log(Var_X)) / 2.0
+            Var_X = np.diag(Var_X)
```

Comment:
It will not. However, there is also the danger of modifying mutable objects; hopefully that is not happening here.
https://docs.python.org/3/faq/programming.html#how-do-i-write-a-function-with-output-parameters-call-by-reference